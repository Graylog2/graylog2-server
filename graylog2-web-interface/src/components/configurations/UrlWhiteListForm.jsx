// @flow strict
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';
import { cloneDeep, debounce } from 'lodash';
import Input from 'components/bootstrap/Input';
import { Select, Icon } from 'components/common';
import { Button, Table } from 'components/graylog';
import FormUtils from 'util/FormsUtils';
import type { Url, WhiteListConfig } from 'stores/configurations/ConfigurationsStore';
import StoreProvider from 'injection/StoreProvider';

const ToolsStore = StoreProvider.getStore('Tools');

type Props = {
  urls: Array<Url>,
  disabled: boolean,
  onUpdate: (config: WhiteListConfig, valid: boolean) => void
};

const UrlWhiteListForm = ({ urls, onUpdate, disabled }: Props) => {
  const literal = 'literal';
  const regex = 'regex';
  const options = [{ value: literal, label: 'Exact match' }, { value: regex, label: 'Regex' }];
  // eslint-disable-next-line prefer-const
  let inputs = {};
  const [config, setConfig] = useState<WhiteListConfig>({ entries: urls, disabled });
  const [validationState, setValidationState] = useState({ errors: [] });

  const _onAdd = (event: Event) => {
    event.preventDefault();
    setConfig({ ...config, entries: [...config.entries, { id: uuid(), title: '', value: '', type: literal }] });
  };

  const _onRemove = (event: MouseEvent, idx: number) => {
    event.preventDefault();
    // eslint-disable-next-line prefer-const
    let stateUpdate = cloneDeep(config);
    const validationUpdate = cloneDeep(validationState);
    validationUpdate.errors[idx] = null;
    setValidationState(validationUpdate);
    stateUpdate.entries.splice(idx, 1);
    setConfig(stateUpdate);
  };

  const validURL = (str: string) => {
    let isValid = true;
    try {
      // eslint-disable-next-line no-unused-vars
      const test = new URL(str);
    } catch (e) {
      isValid = false;
    }
    return isValid;
  };

  const _isFormValid = (): boolean => {
    let isValid = true;
    if (validationState.errors.length > 0
      && validationState.errors.find((el => (el && el.title && el.title.valid) === false
      || (el && el.value && el.value.valid === false)))) {
      isValid = false;
    }
    return isValid;
  };

  const _updateState = (idx: number, type: string, name: string, value: string) => {
    const stateUpdate = cloneDeep(config);
    stateUpdate.entries[idx][name] = value;
    stateUpdate.entries[idx] = { ...stateUpdate.entries[idx], type };
    setConfig(stateUpdate);
  };

  const _updateValidationError = (idx: number, type: string, name: string, result: Object, value: string) => {
    const validationUpdate = cloneDeep(validationState);
    validationUpdate.errors[idx] = { ...validationUpdate.errors[idx], [name]: result };
    setValidationState(validationUpdate);
    _updateState(idx, type, name, value);
  };

  const _validate = (name: string, idx: number, type: string, value: string): void => {
    switch (name) {
      case 'title': {
        const result = value.trim().length <= 0 ? { valid: false } : { valid: true };
        _updateValidationError(idx, type, name, result, value);
      }
        break;
      case 'value':
        if (type === literal) {
          const result = validURL(value) ? { valid: true } : { valid: false };
          _updateValidationError(idx, type, name, result, value);
        } else if (type === regex && value.trim().length > 0) {
          const promise = ToolsStore.testRegexValidity(value);
          promise.then((result) => {
            const res = result.is_valid ? { valid: true } : { valid: false };
            _updateValidationError(idx, type, name, res, value);
          });
        } else {
          const res = { valid: false };
          _updateValidationError(idx, type, name, res, value);
        }
        break;
      default:
        break;
    }
  };

  const debouncedValidate = debounce(_validate, 500);

  const _onInputChange = (event: SyntheticInputEvent<EventTarget>, idx: number, type: string) => {
    if (type === regex) {
      debouncedValidate(event.target.name, idx, type, FormUtils.getValueFromInput(event.target));
    } else {
      _validate(event.target.name, idx, type, FormUtils.getValueFromInput(event.target));
    }
  };

  const _onUpdateType = (idx: number, type: string) => {
    const stateUpdate = cloneDeep(config);
    _validate('value', idx, type, stateUpdate.entries[idx].value);
  };

  const _getErrorMessage = (type: string) => {
    return type === regex ? 'Not a valid Java regular expression' : 'Not a valid URL';
  };

  const _getSummary = () => {
    return (config.entries.map((url, idx) => {
      return (
        <tr key={url.id}>
          <td style={{ verticalAlign: 'middle', textAlign: 'center' }}>{idx + 1}</td>
          <td>
            <Input type="text"
                   id={`title-input${idx}`}
                   ref={(elem) => { inputs[`title${idx}`] = elem; }}
                   help={validationState.errors[idx] && validationState.errors[idx].title && !validationState.errors[idx].title.valid ? 'Required field' : null}
                   name="title"
                   bsStyle={validationState.errors[idx] && validationState.errors[idx].title && !validationState.errors[idx].title.valid ? 'error' : null}
                   onChange={event => _onInputChange(event, idx, url.type)}
                   defaultValue={url.title}
                   required />
          </td>
          <td>
            <Input type="text"
                   id={`value-input${idx}`}
                   ref={(elem) => { inputs[`value${idx}`] = elem; }}
                   help={validationState.errors[idx] && validationState.errors[idx].value && !validationState.errors[idx].value.valid ? _getErrorMessage(url.type) : null}
                   name="value"
                   bsStyle={validationState.errors[idx] && validationState.errors[idx].value && !validationState.errors[idx].value.valid ? 'error' : null}
                   onChange={event => _onInputChange(event, idx, url.type)}
                   defaultValue={url.value}
                   required />
          </td>
          <td>
            <Input id={`url-input-type-${idx}`}
                   required
                   autoFocus>
              <Select clearable={false}
                      options={options}
                      matchProp="label"
                      onChange={option => _onUpdateType(idx, option)}
                      value={url.type} />
            </Input>
          </td>
          <td>
            <Button onClick={event => _onRemove(event, idx)}>
              <Icon name="fa-trash" />
            </Button>
          </td>
        </tr>
      );
    }));
  };

  useEffect(() => {
    const valid = _isFormValid();
    onUpdate(config, valid);
  }, [config]);

  return (
    <>
      <Input type="checkbox"
             id="whitelist-disabled"
             label="Disable Whitelist"
             checked={config.disabled}
             onChange={() => setConfig({ ...config, disabled: !config.disabled })}
             help="Disable the whitelist functionality. Warning: Disabling this option will allow users to enter any URL in Graylog entities, which may pose a security risk." />
      <Button bsSize="sm" onClick={event => _onAdd(event)}>Add Url</Button>
      <Table striped bordered className="top-margin">
        <thead>
          <tr>
            <th>#</th>
            <th>Title</th>
            <th>URL</th>
            <th>Type</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {_getSummary()}
        </tbody>
      </Table>
      <Button bsSize="sm" onClick={event => _onAdd(event)}>Add Url</Button>
    </>
  );
};

UrlWhiteListForm.propTypes = {
  urls: PropTypes.array,
  disabled: PropTypes.bool,
  onUpdate: PropTypes.func,
};

UrlWhiteListForm.defaultProps = {
  urls: [],
  disabled: false,
  onUpdate: () => {},
};

export default UrlWhiteListForm;
