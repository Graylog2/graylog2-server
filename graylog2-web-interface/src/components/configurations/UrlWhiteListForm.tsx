/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';
import { cloneDeep, debounce } from 'lodash';
import styled from 'styled-components';

import Input from 'components/bootstrap/Input';
// Explicit import to fix eslint import/no-cycle
import Select from 'components/common/Select';
import Icon from 'components/common/Icon';
import { Button, Table } from 'components/graylog';
import FormUtils from 'util/FormsUtils';
import type { Url, WhiteListConfig } from 'stores/configurations/ConfigurationsStore';
import StoreProvider from 'injection/StoreProvider';

const ToolsStore = StoreProvider.getStore('Tools');

type Props = {
  urls: Array<Url>,
  disabled: boolean,
  onUpdate: (config: WhiteListConfig, valid: boolean) => void,
};

const StyledTable = styled(Table)`
  margin-top: 10px;
`;

const UrlWhiteListForm = ({ urls, onUpdate, disabled }: Props) => {
  const literal = 'literal';
  const regex = 'regex';
  const options = [{ value: literal, label: 'Exact match' }, { value: regex, label: 'Regex' }];
  // eslint-disable-next-line prefer-const
  let inputs = {};
  const [config, setConfig] = useState<WhiteListConfig>({ entries: urls, disabled });
  const [validationState, setValidationState] = useState({ errors: [] });

  const _onAdd = (event: Event) => {
    event.preventDefault();
    setConfig({ ...config, entries: [...config.entries, { id: uuid(), title: '', value: '', type: literal }] });
  };

  const _onRemove = (event: MouseEvent, idx: number) => {
    event.preventDefault();
    // eslint-disable-next-line prefer-const
    let stateUpdate = cloneDeep(config);
    const validationUpdate = cloneDeep(validationState);

    validationUpdate.errors[idx] = null;
    setValidationState(validationUpdate);
    stateUpdate.entries.splice(idx, 1);
    setConfig(stateUpdate);
  };

  const validURL = (str: string) => {
    let isValid = true;

    try {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const test = new URL(str);
    } catch (e) {
      isValid = false;
    }

    return isValid;
  };

  const _isFormValid = (): boolean => {
    let isValid = true;

    if (validationState.errors.length > 0
      && validationState.errors.find(((el) => (el && el.title && el.title.valid) === false
      || (el && el.value && el.value.valid === false)))) {
      isValid = false;
    }

    return isValid;
  };

  const _updateState = (idx: number, type: string, name: string, value: string) => {
    const stateUpdate = cloneDeep(config);

    stateUpdate.entries[idx][name] = value;
    stateUpdate.entries[idx] = { ...stateUpdate.entries[idx], type };
    setConfig(stateUpdate);
  };

  const _updateValidationError = (idx: number, type: string, name: string, result: Object, value: string) => {
    const validationUpdate = cloneDeep(validationState);

    validationUpdate.errors[idx] = { ...validationUpdate.errors[idx], [name]: result };
    setValidationState(validationUpdate);
    _updateState(idx, type, name, value);
  };

  const _validate = (name: string, idx: number, type: string, value: string): void => {
    switch (name) {
      case 'title': {
        const result = value.trim().length <= 0 ? { valid: false } : { valid: true };

        _updateValidationError(idx, type, name, result, value);
      }

        break;
      case 'value':
        if (type === literal) {
          const result = validURL(value) ? { valid: true } : { valid: false };

          _updateValidationError(idx, type, name, result, value);
        } else if (type === regex && value.trim().length > 0) {
          const promise = ToolsStore.testRegexValidity(value);

          promise.then((result) => {
            const res = result.is_valid ? { valid: true } : { valid: false };

            _updateValidationError(idx, type, name, res, value);
          });
        } else {
          const res = { valid: false };

          _updateValidationError(idx, type, name, res, value);
        }

        break;
      default:
        break;
    }
  };

  const debouncedValidate = debounce(_validate, 500);

  const _onInputChange = (event: React.ChangeEvent<HTMLInputElement>, idx: number, type: string) => {
    if (type === regex) {
      debouncedValidate(event.target.name, idx, type, FormUtils.getValueFromInput(event.target));
    } else {
      _validate(event.target.name, idx, type, FormUtils.getValueFromInput(event.target));
    }
  };

  const _onUpdateType = (idx: number, type: string) => {
    const stateUpdate = cloneDeep(config);

    _validate('value', idx, type, stateUpdate.entries[idx].value);
  };

  const _getErrorMessage = (type: string) => {
    return type === regex ? 'Not a valid Java regular expression' : 'Not a valid URL';
  };

  const _getSummary = () => {
    return (config.entries.map((url, idx) => {
      return (
        <tr key={url.id}>
          <td style={{ verticalAlign: 'middle', textAlign: 'center' }}>{idx + 1}</td>
          <td>
            <Input type="text"
                   id={`title-input${idx}`}
                   ref={(elem) => { inputs[`title${idx}`] = elem; }}
                   help={validationState.errors[idx] && validationState.errors[idx].title && !validationState.errors[idx].title.valid ? 'Required field' : null}
                   name="title"
                   bsStyle={validationState.errors[idx] && validationState.errors[idx].title && !validationState.errors[idx].title.valid ? 'error' : null}
                   onChange={(event) => _onInputChange(event, idx, url.type)}
                   defaultValue={url.title}
                   required />
          </td>
          <td>
            <Input type="text"
                   id={`value-input${idx}`}
                   ref={(elem) => { inputs[`value${idx}`] = elem; }}
                   help={validationState.errors[idx] && validationState.errors[idx].value && !validationState.errors[idx].value.valid ? _getErrorMessage(url.type) : null}
                   name="value"
                   bsStyle={validationState.errors[idx] && validationState.errors[idx].value && !validationState.errors[idx].value.valid ? 'error' : null}
                   onChange={(event) => _onInputChange(event, idx, url.type)}
                   defaultValue={url.value}
                   required />
          </td>
          <td>
            <Input id={`url-input-type-${idx}`}
                   required
                   autoFocus>
              <Select clearable={false}
                      options={options}
                      matchProp="label"
                      placeholder="Select url type"
                      onChange={(option) => _onUpdateType(idx, option)}
                      value={url.type} />
            </Input>
          </td>
          <td>
            <Button onClick={(event) => _onRemove(event, idx)}>
              <Icon name="trash-alt" />
            </Button>
          </td>
        </tr>
      );
    }));
  };

  useEffect(() => {
    const valid = _isFormValid();

    onUpdate(config, valid);
  }, [config]);

  return (
    <>
      <Input type="checkbox"
             id="whitelist-disabled"
             label="Disable Whitelist"
             checked={config.disabled}
             onChange={() => setConfig({ ...config, disabled: !config.disabled })}
             help="Disable the whitelist functionality. Warning: Disabling this option will allow users to enter any URL in Graylog entities, which may pose a security risk." />
      <Button bsSize="sm" onClick={(event) => _onAdd(event)}>Add Url</Button>
      <StyledTable striped bordered>
        <thead>
          <tr>
            <th>#</th>
            <th>Title</th>
            <th>URL</th>
            <th>Type</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {_getSummary()}
        </tbody>
      </StyledTable>
      <Button bsSize="sm" onClick={(event) => _onAdd(event)}>Add Url</Button>
    </>
  );
};

UrlWhiteListForm.propTypes = {
  urls: PropTypes.array,
  disabled: PropTypes.bool,
  onUpdate: PropTypes.func,
};

UrlWhiteListForm.defaultProps = {
  urls: [],
  disabled: false,
  onUpdate: () => {},
};

export default UrlWhiteListForm;
