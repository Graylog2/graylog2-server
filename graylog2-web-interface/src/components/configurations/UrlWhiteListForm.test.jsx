// @flow strict
import React from 'react';
import { mount } from 'wrappedEnzyme';
import 'helpers/mocking/react-dom_mock';
import UrlWhiteListForm from './UrlWhiteListForm';

describe('UrlWhitelistForm', () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation(init => [init, setState]);
  const onUpdate = jest.fn();
  const config = {
    entries: [
      {
        id: 'f7033f1f-d50f-4323-96df-294ede41d951',
        value: 'http://localhost:8080/system/',
        title: 'testam',
        type: 'regex',
      },
      {
        id: '636a2d40-c4c5-40b9-ab3a-48cf7978e9af',
        value: 'http://localhost:8080/system/',
        title: 'test',
        type: 'regex',
      },
      {
        id: 'f28fd891-5f2d-4128-9a94-e97c1ab07a1f',
        value: 'http://localhost:8080/system/',
        title: 'test',
        type: 'literal',
      },
    ],
    disabled: false,
  };


  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('render the UrlWhitelistForm component', () => {
    it('should create new instance', () => {
      wrapper = mount(<UrlWhiteListForm urls={config.entries}
                                        disabled={config.disabled}
                                        onUpdate={onUpdate} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should display Url form list table', () => {
      wrapper = mount(<UrlWhiteListForm urls={config.entries}
                                        disabled={config.disabled}
                                        onUpdate={onUpdate} />);
      expect(wrapper.find('.form-group')).toBeDefined();
      const title = wrapper.find('input#title-input0').at(0);
      expect(title.instance().value).toBe(config.entries[0].title);
    });

    it('should validate and update on input change', () => {
      wrapper = mount(<UrlWhiteListForm urls={config.entries}
                                        disabled={config.disabled}
                                        onUpdate={onUpdate} />);
      const title = wrapper.find('input#title-input0').at(0);
      title.instance().value = 'world';
      title.simulate('change');
      expect(onUpdate).toHaveBeenCalled();
    });

    it('should add a new row to the form', () => {
      wrapper = mount(<UrlWhiteListForm urls={config.entries}
                                        disabled={config.disabled}
                                        onUpdate={onUpdate} />);
      const button = wrapper.find('button').at(0);
      button.simulate('click');
      const listItems = wrapper.find('tbody>tr');
      expect(listItems.length).toBe(config.entries.length + 1);
    });

    it('should delete a row', () => {
      wrapper = mount(<UrlWhiteListForm urls={config.entries}
                                        disabled={config.disabled}
                                        onUpdate={onUpdate} />);
      const deleteButton = wrapper.find('i.fa-trash').at(0);
      deleteButton.simulate('click');
      const listItems = wrapper.find('tbody>tr');
      expect(listItems.length).toBe(config.entries.length - 1);
    });
  });
});
