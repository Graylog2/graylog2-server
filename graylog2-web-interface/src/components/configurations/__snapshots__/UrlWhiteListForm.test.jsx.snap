// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlWhitelistForm render the UrlWhitelistForm component should create new instance 1`] = `
Array [
  <div
    className="form-group"
  >
    <span>
      <div
        className="checkbox"
      >
        <label
          title=""
        >
          <input
            checked={false}
            disabled={false}
            label="Disable Whitelist"
            name="whitelist-disabled"
            onChange={[Function]}
            placeholder=""
            type="checkbox"
          />
          Disable Whitelist
        </label>
      </div>
      <span
        className="help-block"
      >
        Disable the whitelist functionality. Warning: Disabling this option will allow users to enter any URL in Graylog entities, which may pose a security risk.
      </span>
    </span>
  </div>,
  <button
    className="btn btn-sm btn-default"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add Url
  </button>,
  <table
    className="top-margin table table-striped table-bordered"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Title
        </th>
        <th>
          URL
        </th>
        <th>
          Type
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style={
            Object {
              "textAlign": "center",
              "verticalAlign": "middle",
            }
          }
        >
          1
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <input
                className="form-control"
                defaultValue="testam"
                id="title-input0"
                label=""
                name="title"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="text"
              />
            </span>
          </div>
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <input
                className="form-control"
                defaultValue="http://localhost:8080/system/"
                id="value-input0"
                label=""
                name="value"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="text"
              />
            </span>
          </div>
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <div
                className="css-1pcexqc-container"
                onKeyDown={[Function]}
              >
                <div
                  className="css-mxcylq-control"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <div
                    className="css-1hio1c"
                  >
                    <div
                      className="css-13zpnff-singleValue"
                    >
                      Regex
                    </div>
                    <div
                      className="css-1g6gooi"
                    >
                      <div
                        className=""
                        style={
                          Object {
                            "display": "inline-block",
                          }
                        }
                      >
                        <input
                          aria-autocomplete="list"
                          autoCapitalize="none"
                          autoComplete="off"
                          autoCorrect="off"
                          disabled={false}
                          id="react-select-2-input"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          spellCheck="false"
                          style={
                            Object {
                              "background": 0,
                              "border": 0,
                              "boxSizing": "content-box",
                              "color": "inherit",
                              "fontSize": "inherit",
                              "label": "input",
                              "opacity": 1,
                              "outline": 0,
                              "padding": 0,
                              "width": "1px",
                            }
                          }
                          tabIndex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style={
                            Object {
                              "height": 0,
                              "left": 0,
                              "overflow": "scroll",
                              "position": "absolute",
                              "top": 0,
                              "visibility": "hidden",
                              "whiteSpace": "pre",
                            }
                          }
                        >
                          
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    className="css-1wy0on6"
                  >
                    <div
                      aria-hidden="true"
                      onMouseDown={[Function]}
                      onTouchEnd={[Function]}
                      style={
                        Object {
                          ":hover": Object {
                            "color": "hsl(0, 0%, 60%)",
                          },
                          "boxSizing": "border-box",
                          "color": "hsl(0, 0%, 80%)",
                          "display": "flex",
                          "fontSize": "150%",
                          "label": "indicatorContainer",
                          "marginRight": "1rem",
                          "padding": "0px",
                          "transform": false,
                          "transition": "color 150ms",
                        }
                      }
                    >
                      <i
                        className="fa fa-caret-down"
                      />
                    </div>
                  </div>
                </div>
              </div>
              
            </span>
          </div>
        </td>
        <td>
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-trash"
            />
          </button>
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "textAlign": "center",
              "verticalAlign": "middle",
            }
          }
        >
          2
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <input
                className="form-control"
                defaultValue="test"
                id="title-input1"
                label=""
                name="title"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="text"
              />
            </span>
          </div>
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <input
                className="form-control"
                defaultValue="http://localhost:8080/system/"
                id="value-input1"
                label=""
                name="value"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="text"
              />
            </span>
          </div>
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <div
                className="css-1pcexqc-container"
                onKeyDown={[Function]}
              >
                <div
                  className="css-mxcylq-control"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <div
                    className="css-1hio1c"
                  >
                    <div
                      className="css-13zpnff-singleValue"
                    >
                      Regex
                    </div>
                    <div
                      className="css-1g6gooi"
                    >
                      <div
                        className=""
                        style={
                          Object {
                            "display": "inline-block",
                          }
                        }
                      >
                        <input
                          aria-autocomplete="list"
                          autoCapitalize="none"
                          autoComplete="off"
                          autoCorrect="off"
                          disabled={false}
                          id="react-select-3-input"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          spellCheck="false"
                          style={
                            Object {
                              "background": 0,
                              "border": 0,
                              "boxSizing": "content-box",
                              "color": "inherit",
                              "fontSize": "inherit",
                              "label": "input",
                              "opacity": 1,
                              "outline": 0,
                              "padding": 0,
                              "width": "1px",
                            }
                          }
                          tabIndex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style={
                            Object {
                              "height": 0,
                              "left": 0,
                              "overflow": "scroll",
                              "position": "absolute",
                              "top": 0,
                              "visibility": "hidden",
                              "whiteSpace": "pre",
                            }
                          }
                        >
                          
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    className="css-1wy0on6"
                  >
                    <div
                      aria-hidden="true"
                      onMouseDown={[Function]}
                      onTouchEnd={[Function]}
                      style={
                        Object {
                          ":hover": Object {
                            "color": "hsl(0, 0%, 60%)",
                          },
                          "boxSizing": "border-box",
                          "color": "hsl(0, 0%, 80%)",
                          "display": "flex",
                          "fontSize": "150%",
                          "label": "indicatorContainer",
                          "marginRight": "1rem",
                          "padding": "0px",
                          "transform": false,
                          "transition": "color 150ms",
                        }
                      }
                    >
                      <i
                        className="fa fa-caret-down"
                      />
                    </div>
                  </div>
                </div>
              </div>
              
            </span>
          </div>
        </td>
        <td>
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-trash"
            />
          </button>
        </td>
      </tr>
      <tr>
        <td
          style={
            Object {
              "textAlign": "center",
              "verticalAlign": "middle",
            }
          }
        >
          3
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <input
                className="form-control"
                defaultValue="test"
                id="title-input2"
                label=""
                name="title"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="text"
              />
            </span>
          </div>
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <input
                className="form-control"
                defaultValue="http://localhost:8080/system/"
                id="value-input2"
                label=""
                name="value"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="text"
              />
            </span>
          </div>
        </td>
        <td>
          <div
            className="form-group"
          >
            
            <span>
              <div
                className="css-1pcexqc-container"
                onKeyDown={[Function]}
              >
                <div
                  className="css-mxcylq-control"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <div
                    className="css-1hio1c"
                  >
                    <div
                      className="css-13zpnff-singleValue"
                    >
                      Exact match
                    </div>
                    <div
                      className="css-1g6gooi"
                    >
                      <div
                        className=""
                        style={
                          Object {
                            "display": "inline-block",
                          }
                        }
                      >
                        <input
                          aria-autocomplete="list"
                          autoCapitalize="none"
                          autoComplete="off"
                          autoCorrect="off"
                          disabled={false}
                          id="react-select-4-input"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          spellCheck="false"
                          style={
                            Object {
                              "background": 0,
                              "border": 0,
                              "boxSizing": "content-box",
                              "color": "inherit",
                              "fontSize": "inherit",
                              "label": "input",
                              "opacity": 1,
                              "outline": 0,
                              "padding": 0,
                              "width": "1px",
                            }
                          }
                          tabIndex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style={
                            Object {
                              "height": 0,
                              "left": 0,
                              "overflow": "scroll",
                              "position": "absolute",
                              "top": 0,
                              "visibility": "hidden",
                              "whiteSpace": "pre",
                            }
                          }
                        >
                          
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    className="css-1wy0on6"
                  >
                    <div
                      aria-hidden="true"
                      onMouseDown={[Function]}
                      onTouchEnd={[Function]}
                      style={
                        Object {
                          ":hover": Object {
                            "color": "hsl(0, 0%, 60%)",
                          },
                          "boxSizing": "border-box",
                          "color": "hsl(0, 0%, 80%)",
                          "display": "flex",
                          "fontSize": "150%",
                          "label": "indicatorContainer",
                          "marginRight": "1rem",
                          "padding": "0px",
                          "transform": false,
                          "transition": "color 150ms",
                        }
                      }
                    >
                      <i
                        className="fa fa-caret-down"
                      />
                    </div>
                  </div>
                </div>
              </div>
              
            </span>
          </div>
        </td>
        <td>
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-trash"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>,
  <button
    className="btn btn-sm btn-default"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add Url
  </button>,
]
`;
